src/Core/Checkout/Customer/SalesChannel/AccountService.php

```diff
@@ -179,46 +182,94 @@ private function loginByCustomer(CustomerEntity $customer, SalesChannelContext $
     {
         $criteria = new Criteria([$id]);
-        $customer = $this->fetchCustomer($criteria, $context);
 
+        $customer = $this->fetchCustomer($criteria, $context, true);
         if ($customer === null) {
             throw new CustomerNotFoundByIdException($id);
         }
 
         return $customer;
     }
 
-    private function fetchCustomer(Criteria $criteria, SalesChannelContext $context): ?CustomerEntity
+    /**
+     * Add only filters to the $criteria for values which have an index in the database, e.g. id, or email,
+     * the rest should be done via PHP. This method filters for the standard customer related constraints
+     * like active or the sales channel assignment.
+     *
+     * @throws BadCredentialsException
+     */
+    private function fetchCustomer(Criteria $criteria, SalesChannelContext $context, bool $includeGuest = false): ?CustomerEntity
     {
-        $criteria->addFilter(new MultiFilter(MultiFilter::CONNECTION_OR, [
-            new EqualsFilter('boundSalesChannelId', null),
-            new EqualsFilter('boundSalesChannelId', $context->getSalesChannel()->getId()),
-        ]));
+        $criteria->setTitle('account-service');
+
+        $result = $this->customerRepository->search($criteria, $context->getContext());
+        $result = $result->filter(static function (CustomerEntity $customer) use ($includeGuest, $context) {
+            // Skip not active users
+            if (!$customer->getActive()) {
+                $isConfirmed = !$customer->getDoubleOptInRegistration() || $customer->getDoubleOptInConfirmDate();
+
+                // Customers with double optin will be active by default starting at Shopware 6.5.0.0,
+                // remove complete if statement and always return null
+                if (Feature::isActive('v6.5.0.0') || $isConfirmed) {
+                    return null;
+                }
+            }
+
+            // Skip guest if not required
+            if (!$includeGuest && $customer->getGuest()) {
+                return null;
+            }
+
+            // If not bound, we still need to consider it
+            if ($customer->getBoundSalesChannelId() === null) {
+                return true;
+            }
+
+            // It is bound, but not to the current one. Skip it
+            if ($customer->getBoundSalesChannelId() !== $context->getSalesChannel()->getId()) {
+                return null;
+            }
+
+            return true;
+        });
+
+        // There can be multiple guest accounts, here we only want the latest one
+        if ($includeGuest) {
+            if ($result->count() > 1) {
+                $result->sort(function (CustomerEntity $a, CustomerEntity $b) {
+                    return $a->getCreatedAt() <=> $b->getCreatedAt();
+                });
+            }
+
+            return $result->first();
+        }
+
+        if ($result->count() !== 1) {
```

Comment:
i'm not sure if this is too strict, as on DB level there is no check that there is only one none guest customer per sales channel, and i guess there are installations out there where the data is corrupted in that way, even as it is not possible to create such data from shopware itself

I would suggest that we simply always return the latest account if there are more then one and don't need to check for the guest flag there. As we would then always sort by the creationDate we can maybe add the sorting to the criteria, so we don't have to do it on php side. My assumption would be that mysql does the sorting after it filtered down the result set, so there would only be a minimal performance impact compared to not sorting over mysql

Reply:
Mh yes, I agree that like it is solved is not really ideal. I had the problem in the past, that the records in the database were corrupted and there were multiple records with the same mail.

I wanted to handle this case, but this does not really helps, in this case as well, in fact it might be more confusing. So I implemented your suggestion. This is a problem for another time :-)